### DISCUSSION TITLE
Yt-dlp not working in production specially inside the laravel queue.

### DISCUSSION THREAD

**Discussion by subhamghimire** [norkunas/youtube-dl-php#236] [Q&A]
Hello everyone,

Here's the code inside a Laravel job, which downloads YouTube audio using yt-dlp. It works perfectly fine locally. In fact, when I run certain commands like $collection->getVideos() in the production server's tinker, it retrieves the audio file. works 100%.

However, when I execute the same working code inside the job, $collection is empty.
```
$title = "";
$videoId = "";
$path = Storage::disk('public')->path('tracks');

$yt = new YoutubeDl();
$yt->setBinPath('/usr/local/bin/yt-dlp');

$collection = $yt->download(
    Options::create()
        ->downloadPath($path)
        ->extractAudio(true)
        ->audioFormat('mp3')
        ->audioQuality('0')
        ->output('%(id)s.%(ext)s')
        ->url($this->link)
);

foreach ($collection->getVideos() as $video) {
    if ($video->getError() !== null) {
        throw new \Exception("Error while processing video");
    } else {
        $title = $video->getTitle();
        $videoId = $video->getId();
    }
}

$audioUrl = $videoId . '.mp3';
```

**Comment by norkunas** [comment]
Hi, you could try to enable process debugging to see what's going on:

```php
$yt->debug(function ($type, $buffer) {
    echo "$type: $buffer;"
});
```

**Comment by subhamghimire** [comment]
Okay @norkunas I'm trying this now and will get back to you asap

**Comment by subhamghimire** [comment]
@norkunas It's not echo anything. Please help me

**Comment by subhamghimire** [comment]
@norkunas When this job is run its immediately run in prod under 1 second. So I'm confused.
![image](https://github.com/norkunas/youtube-dl-php/assets/26000785/9e96b48f-74a5-423a-9a2b-e63eeeb38f92)

**Reply by norkunas** [reply]
I'm not sure then what's the cause, I think you should use real debugger like xdebug

**Reply by norkunas** [reply]
Also - did you ensure that the file was not downloaded yet? Because it won't download again if it exists already

**Reply by norkunas** [reply]
Try something like this in your background task:
```
$process = new \Symfony\Component\Process\Process([
    '/usr/local/bin/yt-dlp',
    '--ignore-config',
    '--ignore-errors',
    '--write-info-json',
    '--output=/path/to/your-downloads/%(id)s.%(ext)s',
    '--extract-audio',
    '--audio-format=mp3',
    '--audio-quality=0',
    'https://www.youtube.com/watch?v=oDAw7vW7H0c'
]);
$process->setTimeout(null);
$process->mustRun(function ($type, $buffer) {
    echo "$type: $buffer";
});
```

we'll see if this is something related to the library or at the lower level

**Reply by subhamghimire** [reply]
Okay @norkunas. I'll try & let you know.

**Comment by subhamghimire** [comment]
Yes. I tried with new youtube video @norkunas

**Comment by subhamghimire** [comment]
@norkunas I have tried the above code. And got this error:
```The command "'/usr/local/bin/yt-dlp' '--ignore-config' '--ignore-errors' '--write-info-json' '--output=/var/www/storage/app/public/efs/tracks/%(id)s.%(ext)s' '--extract-audio' '--audio-format=mp3' '--audio-quality=0' 'https://www.youtube.com/watch?v=MK24OO1zgYc'" failed. Exit Code: 127(Command not found) Working directory: /var/www Output: ================ Error Output: ================ sh: 1: exec: /usr/local/bin/yt-dlp: not found```

But, yt-dlp file exists in the necessary directory

**Reply by norkunas** [reply]
So now that the problem is localized, and we know that it's not the library fault, you should try to debug why the yt-dlp binary is not reachable with subprocess in your bg task :)

**Reply by subhamghimire** [reply]
Yes not library fault, I'll check if there's permissions issue or something else. Would you recommend what can I  do next to deal this problem ?

**Reply by norkunas** [reply]
I'd try to check if it properly locates with:
```php
exec('which yt-dlp')
```
in your bg task

**Reply by subhamghimire** [reply]
@norkunas I got empty response when execute above code inside the background job

**Reply by norkunas** [reply]
that means for some reason php cannot see your yt-dlp at all, check your php config, etc.

**Reply by subhamghimire** [reply]
@norkunas Can I place yt-dlp file to somewhere else like public folder and access it ? Does it work even if many users are using yt-dlp same time ?

**Reply by norkunas** [reply]
It's up to you where you store your binary, just make sure that it is accessible by your php process

**Reply by subhamghimire** [reply]
@norkunas Can I store yt-dlp binary file directly in the repository ? Does it affects project in the like security risks, problem when executing simultaneously or make slow ?

**Reply by norkunas** [reply]
> Can I store yt-dlp binary file directly in the repository ?

I suggest to use package managers, so then you can update it regularly

> Does it affects project in the like security risks

No, if you don't give direct access to it from outside

> problem when executing simultaneously or make slow ?

that depends on your server

**Reply by subhamghimire** [reply]
@norkunas Could you please guide me on how to use package manager. I didn't get it how binary file would be available ?

**Reply by norkunas** [reply]
https://github.com/yt-dlp/yt-dlp/wiki/Installation#third-party-package-managers or use releases from https://github.com/yt-dlp/yt-dlp?tab=readme-ov-file#release-files and then you'll be able to use `yt-dlp -U` to update

**Reply by subhamghimire** [reply]
@norkunas What would be the bin path. Do I need to set bin path(setBinPath) if I used third party package manager ?
```sudo add-apt-repository ppa:tomtomtom/yt-dlp    # Add ppa repo to apt
sudo apt update                                 # Update package list
sudo apt install yt-dlp  ```

**Reply by norkunas** [reply]
I'm not sure, just run `which yt-dlp`

**Reply by subhamghimire** [reply]
Okay. Thank you @norkunas
